lib/tests/streamlit/components_test.py

```diff
@@ -26,123 +28,129 @@
+    def test_only_url(self):
+        """Test __init__ when only url is provided."""
+        instance = CustomComponent(None, URL)
+        self.assertEqual(URL, instance.url)
+        self.assertIsNone(instance.path)
+        self.assertIsNone(instance._custom_wrapper)
 
+    def test_path_and_url(self):
+        """Test __init__ when both path and url are provided."""
+        with pytest.raises(StreamlitAPIException) as exception_message:
+            CustomComponent(PATH, URL)
+        self.assertEqual(
+            "Either 'path' or 'url' must be set, but not both.",
+            str(exception_message.value),
+        )
 
-class ComponentRequestHandlerTest(tornado.testing.AsyncHTTPTestCase):
-    """Tests the /plugin endpoint"""
+    def test_no_path_and_no_url(self):
+        """Test __init__ when neither path nor url are provided."""
+        with pytest.raises(StreamlitAPIException) as exception_message:
+            CustomComponent(None, None)
+        self.assertEqual(
+            "Either 'path' or 'url' must be set, but not both.",
+            str(exception_message.value),
+        )
+
+
+class RegisterComponentTest(DeltaGeneratorTestCase):
+    """Test registration of custom component."""
+
+    def setUp(self):
+        super().setUp()
+        TestComponent = declare_component(url=URL)
+        register_component("test_component", TestComponent)
+
+    def test_st_binding(self):
+        """Test that component has been attached to st namespace"""
+        self.assertTrue(hasattr(streamlit.components.st, "test_component"))
+
+    def test_dg_binding(self):
+        """Test that component has been attached to DeltaGenerator"""
+        self.assertTrue(hasattr(streamlit.components.DeltaGenerator, "test_component"))
 
-    # TODO!
+    def test_only_json_args(self):
+        """Test that component with only json args is marshalled correctly."""
+        streamlit.test_component(foo="bar")
+        proto = self.get_delta_from_queue().new_element.component_instance
+        self.assertEqual(json.dumps({"foo": "bar"}), proto.args_json)
+        self.assertEqual("[]", str(proto.args_dataframe))
 
-    def get_app(self):
-        self.registry = ComponentRegistry()
-        return tornado.web.Application(
-            [("/plugin/(.*)", ComponentRequestHandler, dict(registry=self.registry))]
+    def test_only_df_args(self):
+        """Test that component with only dataframe args is marshalled correctly."""
+        raw_data = {
+            "First Name": ["Jason", "Molly"],
+            "Last Name": ["Miller", "Jacobson"],
+            "Age": [42, 52],
+        }
+        df = pd.DataFrame(raw_data, columns=["First Name", "Last Name", "Age"])
+        streamlit.test_component(df=df)
+        proto = self.get_delta_from_queue().new_element.component_instance
+        self.assertEqual("{}", proto.args_json)
+        # (HK) TODO: Add assertEqual check for Apache Arrow pybytes.
+        self.assertIsNotNone(proto.args_dataframe)
```

Comment:
Can you add another test that ensures that objects that are `is_dataframe_compatible()` but not `is_dataframe_like()` do not get serialized as dataframes? (Basically, let's test that passing a regular array or dict as an argument to a component results in a JSON array or dict instead of a dataframe).